src/sentry/ratelimits/sliding_windows.py

```diff
@@ -0,0 +1,209 @@
+
+
+@dataclass(frozen=True)
+class RequestedQuota:
+    # A string that all redis state is prefixed with. For example
+    # `sentry-string-indexer:123` where 123 is an organization id.
+    #
+    # Note: You cannot control the redis sharding this way, so curly braces are
+    # forbidden.
+    prefix: str
+
+    # How much of each quota's limit is requested
+    requested: int
+
+    # Which quotas to check against. The requested amount must "fit" into all
+    # quotas.
+    quotas: Sequence[Quota]
+
+
+@dataclass(frozen=True)
+class GrantedQuota:
+    # The prefix from RequestedQuota
+    prefix: str
+
+    # How much of RequestedQuota.requested can actually be used.
+    granted: int
+
+
+class SlidingWindowRateLimiter(Service):
+    def __init__(self, **options: Any) -> None:
+        pass
+
+    def check_within_quotas(
+        self, requests: Sequence[RequestedQuota], timestamp: Optional[int] = None
+    ) -> Tuple[int, Sequence[GrantedQuota]]:
+        """
+        Given a set of quotas requests and limits, compute how much quota could
+        be consumed.
+        """
+        raise NotImplementedError()
+
+    def use_quotas(
+        self, requests: Sequence[RequestedQuota], grants: Sequence[GrantedQuota], timestamp: int
+    ) -> None:
+        """
+        Given a set of requests and the corresponding return values from
+        `check_within_quotas`, consume the quotas.
+
+        This is a separate method because 1) the redis backend is not strongly
+        consistent anyway 2) the metrics string indexer may not want to
+        increment quotas if it crashes halfway through the database write. In
+        other words, it may want to first check quotas, execute the writes,
+        then apply the quotas.
+        """
+        raise NotImplementedError()
+
+    def check_and_use_quotas(
+        self, requests: Sequence[RequestedQuota], timestamp: Optional[int] = None
+    ) -> Sequence[GrantedQuota]:
+        """
+        Check the quota requests in Redis and consume the quota in one go.
+        """
+        timestamp, grants = self.check_within_quotas(requests, timestamp)
+        self.use_quotas(requests, grants, timestamp)
+        return grants
+
+
+class RedisSlidingWindowRateLimiter(SlidingWindowRateLimiter):
+    def __init__(self, **options: Any) -> None:
+        cluster_key = options.get("cluster", "default")
+        self.client = redis.redis_clusters.get(cluster_key)
+        super().__init__(**options)
+
+    def validate(self) -> None:
+        try:
+            self.client.ping()
+        except Exception as e:
+            raise InvalidConfiguration(str(e))
+
+    def _build_redis_key_raw(self, prefix: str, window: int, granularity: int, granule: int) -> str:
+        return f"sliding-window-rate-limit:{prefix}:{window}:{granularity}:{granule}"
+
+    def _build_redis_key(self, request: RequestedQuota, quota: Quota, granule: int) -> str:
+        return self._build_redis_key_raw(
```

Comment:
Was there any reason to have a key for each granule vss having a set per `prefif:window`  and have each granule represented as an element of the set ?

Reply:
because keys have TTL and set elements don't. yes this increases memory usage due to duplicated keys, but i don't think it's a big problem, we also have the same setup in relay.

I see that the concurrent rate limiter has its own mechanism for implementing TTL inside of sets, but there we need to determine set length/cardinality, and here we don't